hamilton/io/fingerprinting.py

```diff
@@ -0,0 +1,117 @@
+import base64
+import functools
+import hashlib
+import logging
+from collections.abc import Mapping, Sequence, Set
+
+from hamilton.experimental import h_databackends
+
+logger = logging.getLogger(__name__)
+
+
+MAX_DEPTH = 4
+
+
+def _compact_hash(digest: bytes) -> str:
+    """Compact the hash to a string that's safe to pass around.
+
+    NOTE this is particularly relevant for th Hamilton UI and
+    passing hashes/fingerprints through web services.
+    """
+    return base64.urlsafe_b64encode(digest).decode()
+
+
+@functools.singledispatch
+def hash_value(obj, *args, depth=0, **kwargs) -> str:
+    """Fingerprinting strategy that computes a hash of the
+    full Python object.
+
+    The default case hashes the `__dict__` attribute of the
+    object (recursive).
+    """
+    if hasattr(obj, "__dict__") and depth < MAX_DEPTH:
+        depth += 1
+        return hash_value(obj.__dict__, depth=depth)
+
+    hash_object = hashlib.md5("<unhashable>".encode())
+    return _compact_hash(hash_object.digest())
+
+
+@hash_value.register(str)
+@hash_value.register(int)
+@hash_value.register(float)
+@hash_value.register(bool)
+@hash_value.register(bytes)
+def hash_primitive(obj, *args, **kwargs) -> str:
+    """Convert the primitive to a string and hash it"""
+    hash_object = hashlib.md5(str(obj).encode())
+    return _compact_hash(hash_object.digest())
+
+
+@hash_value.register(Sequence)
+def hash_sequence(obj, *, sort: bool = False, **kwargs) -> str:
+    """Hash each object of the sequence.
+
+    Orders matters for the hash since orders matters in a sequence.
+    """
+    if sort:
+        obj = dict(sorted(obj.items()))
+
+    hash_object = hashlib.sha224()
+    for elem in obj:
+        hash_object.update(hash_value(elem).encode())
+
+    return _compact_hash(hash_object.digest())
+
+
+@hash_value.register(Mapping)
+def hash_mapping(obj, *, sort: bool = False, **kwargs) -> str:
+    """Hash each key then its value.
+
+    The mapping is always sorted first because order shouldn't matter
+    in a mapping.
+
+    NOTE this may clash with Python dictionary ordering since >=3.7
```

Comment:
I clarified the note. Here's the case:

```python
foo = {"key": 3, "key2": 13}
bar = {"key2": 13, "key": 3}
```

- Before Python 3.7, dictionaries don't store insertion order. To create a unique hash of a dictionary, it makes sense to sort keys and hash them. Then, `foo` and `bar` have the same hash.

- After Python 3.7 (all Hamilton support-versions), dictionaries store insertion order. In *some* cases, the users might expect `foo` to be unequal to `bar` because they have different insertion order. 

The note was to indicate the "unique hash" produced by `hash_mapping` isn't congruent with the definition of a dictionary since Python 3.7. Nonetheless, I think sorting keys and assuming order doesn't matter remains a better default.

Users can register a function for `collections.ordereddict` if they want to